README
NAME
    Alien::Tidyp - Building, finding and using tidyp library -
    <http://www.tidyp.com>

VERSION
    Version 1.4.6 of Alien::Tidyp uses *tidyp* sources v1.04.

SYNOPSIS
    IMPORTANT: This module is not a perl binding for *tidyp* library; it is
    just a helper module. The real perl binding is implemented by e.g
    HTML::Tidy module, which is able to use Alien::Tidyp to locate *tidyp*
    library on your system (or build it from source codes).

    Alien::Tidyp tries (in given order) during its installation:

    *   Locate an already installed *tidyp* + ask user whether to use the
        already installed *tidyp* or whether to build *tidyp* from sources

    *   Via env variable TIDYP_DIR you can specify where the build script
        should look for the already installed *tidyp* (directories
        $TIDYP_DIR/lib and $TIDYP_DIR/include/tidyp are expected to exist)

    *   When not using the already installed *tidyp* build process continues
        with the following steps

    *   Download *tidyp* source code tarball

    *   Build *tidyp* binaries from source codes (note: static libraries are
        build in this case)

    *   Install binaries and dev files (*.h, *.a) into *share* directory of
        Alien::Tidyp distribution - *share* directory is usually something
        like this: /usr/lib/perl5/site_perl/5.10/auto/share/dist/Alien-Tidyp

    Later on you can use Alien::Tidyp in your module that needs to link with
    *tidyp* like this:

     # Sample Makefile.pl
     use ExtUtils::MakeMaker;
     use Alien::Tidyp;

     WriteMakefile(
       NAME         => 'Any::Tidyp::Module',
       VERSION_FROM => 'lib/Any/Tidyp/Module.pm',
       LIBS         => Alien::Tidyp->config('LIBS'),
       INC          => Alien::Tidyp->config('INC'),
       # + additional params
     );

    IMPORTANT: As Alien::Tidyp builds static libraries the modules using
    Alien::Tidyp (e.g. HTML::Tidy) need to have Alien::Tidyp just for
    building, not for later use. In other words Alien:Tidyp is just "build
    dependency" not "run-time dependency".

Build.PL options
  --srctarball=<url_or_filename>
    This option might come handy if you are not connected to the Internet.
    You can use it like:

     Build.PL --srctarball=/path/to/file/tidyp-1.04.tar.gz

    or:

     Build.PL --srctarball=http://any.server.com/path/to/tidyp-1.04.tar.gz

    IMPORTANT: The file should always be exactly the same source code
    tarball as specified in the end of Build.PL - see source code (SHA1
    checksum of the tarball is checked).

METHODS
  config()
    This function is the main public interface to this module.

     Alien::Tidyp->config('LIBS');

    Returns a string like: '-L/path/to/tidyp/dir/lib -ltidyp'

     Alien::Tidyp->config('INC');

    Returns a string like: '-I/path/to/tidyp/dir/include/tidyp'

     Alien::Tidyp->config('PREFIX');

    Returns a string like: '/path/to/tidyp/dir' (note: if using the already
    installed tidyp config('PREFIX') returns undef)

AUTHOR
    KMX, <kmx at cpan.org>

BUGS
    Please report any bugs or feature requests to <bug-Alien-Tidyp at
    rt.cpan.org>, or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Alien-Tidyp>.

LICENSE AND COPYRIGHT
    Please notice that the source code of tidyp library has a different
    license than module itself.

  Alien::Tidyp perl module
    Copyright (c) 2010 KMX.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.

  Source code of tidyp library
    Copyright (c) 1998-2003 World Wide Web Consortium (Massachusetts
    Institute of Technology, European Research Consortium for Informatics
    and Mathematics, Keio University). All Rights Reserved.

    This software and documentation is provided "as is," and the copyright
    holders and contributing author(s) make no representations or
    warranties, express or implied, including but not limited to, warranties
    of merchantability or fitness for any particular purpose or that the use
    of the software or documentation will not infringe any third party
    patents, copyrights, trademarks or other rights.

    The copyright holders and contributing author(s) will not be held liable
    for any direct, indirect, special or consequential damages arising out
    of any use of the software or documentation, even if advised of the
    possibility of such damage.

    Permission is hereby granted to use, copy, modify, and distribute this
    source code, or portions hereof, documentation and executables, for any
    purpose, without fee, subject to the following restrictions:

    1. The origin of this source code must not be misrepresented.

    2. Altered versions must be plainly marked as such and must not be
    misrepresented as being the original source.

    3. This Copyright notice may not be removed or altered from any source
    or altered source distribution.

    The copyright holders and contributing author(s) specifically permit,
    without fee, and encourage the use of this source code as a component
    for supporting the Hypertext Markup Language in commercial products. If
    you use this source code in a product, acknowledgment is not required
    but would be appreciated.



Changes
Revision history for Perl module Alien::Tidyp

1.4.7 2011-05-31
    - fixed http://www.cpantesters.org/cpan/report/b74b5c2c-6fa9-1014-aa8f-4b9555dfede8 (force nmake for MSVC build)
    - fixed http://www.cpantesters.org/cpan/report/138b45f2-4b6f-11e0-afaf-8138785ebe45 (ar not in PATH on solaris)

1.4.6 2011-03-10
    - fixed http://www.cpantesters.org/cpan/report/8446953a-4287-11e0-bace-65d4790d71a6 (devnull)
    - fixed http://rt.cpan.org/Ticket/Display.html?id=66382 (MacOSX 10.6.6 failure)

1.4.5 2010-11-26
    - alternative dowload location (too many cpan testers fail to download tarball from github)
    - config.guess + config.sub hack (DragonFly BSD failure)

1.4.4 2010-11-05
    - more error info in case of sha1 checksum failure
    - cosmetic POD changes
    - new option: Build.PL --srctarball=<url_or_filename>

1.4.3 2010-09-25
    - using File::Path legacy functions (works with File::Path <2.0 e.g. RHEL5)

1.4.2 2010-09-23
    - fixing cygwin cpan testers build failure (forcing 'make')
    - deleting the old content of sharedir during install
    - sending all build messages to stderr

1.4.1 2010-09-21
    - tidyp v1.04

1.2.8 2010-08-18
    - fix for 64bit Linux (forcing -fPIC)
    - using -L/path/to instead of -L'/path/to' (EU::MM does not like it)

1.2.7 2010-07-11
    - win32/gcc CPAN Tester failure fix
    - small changes in win32/msvc makefile

1.2.6 2010-07-09
    - windows-msvc support added (both 32/64bit)

1.2.5 2010-07-07
    - linux/amd64 test failure fixed
    - support for env variable TIDYP_DIR introduced

1.2.4 2010-07-06
    - added detection of already installed tidyp
    - leaving out ranlib on mingw build (fixing CPAN Testers failures)

1.2.3 2010-05-11
    - tidyp source not directly include but downloaded before build
    - Win32 dmake support
    - UNIX systems now use ./configure + make

1.2.1 2010-05-10
    - tidyp 1.02 sources

1.0.1 2010-04-25
    - tidyp 1.0 sources
    - gmake as a fallback on solaris reverted

0.99.7 2010-04-06
    - Fix for no_index (another try)
    - Use gmake as a fallback on solaris

0.99.6 2010-04-06
    - Better GNU make detection (solaris, netbsd CPAN tester failures)
    - Fix for no_index

0.99.5 2010-04-04
    - Fixing FreeBSD CPAN tester failures
    - Fixing Win32 CPAN tester failures
    - Build.PL cosmetics
    - Quoting tidyp license

0.99.4 2010-04-02
    - The first CPAN release

0.99.3 2010-03-01
    - Initial version by kmx



Build.PL
use strict;
use warnings;

use lib 'inc';

my %platforms =(
  # Unix = default, thus not listing all UNIX like systems
  MSWin32 => 'Windows',
);
my $package = 'My::Builder::' . ($platforms{$^O} || 'Unix');
print STDERR "Gonna use '$package' class ...\n";
eval "require $package" or die "Require '$package' failed: $@\n";

my $builder = $package->new(
  module_name        => 'Alien::Tidyp',
  dist_version_from  => 'lib/Alien/Tidyp.pm',
  dist_author        => 'KMX, kmx@cpan.org',
  license            => 'perl',
  requires => {
    'perl'              => '5.008',
    'File::ShareDir'    => 0,
    'File::Spec'        => 0,
  },
  configure_requires => {
    'perl'              => '5.008',
    'File::ShareDir'    => 0,
    'File::Spec'        => 0,
    'File::Path'        => 0,
    'File::Temp'        => 0,
    'ExtUtils::CBuilder'=> 0,
    'ExtUtils::Command' => 0,
    'ExtUtils::Liblist' => 0,
    'File::Fetch'       => 0,
    'Digest::SHA'       => 0,
    'Archive::Extract'  => 0,
  },
  build_requires => {
    'perl'              => '5.008',
    'Test::More'        => 0,
    'File::ShareDir'    => 0,
    'File::Spec'        => 0,
    'File::Path'        => 0,
    'File::Temp'        => 0,
    'ExtUtils::CBuilder'=> 0,
    'ExtUtils::Command' => 0,
    'ExtUtils::Liblist' => 0,
    'File::Fetch'       => 0,
    'Digest::SHA'       => 0,
    'Archive::Extract'  => 0,
  },
  share_dir          => 'sharedir',
  add_to_cleanup     => [ 'Alien-Tidyp-*', 'build_done' ],
  create_readme      => 1,
  meta_merge => {
    resources  => {
      bugtracker => 'http://rt.cpan.org/NoAuth/Bugs.html?Dist=Alien-Tidyp',
      repository => 'http://github.com/kmx/alien-tidyp',
    }
  },
  get_options => { srctarball =>  { type  => '=s' } },  # option: Build.PL --srctarball=<url_or_filename>
);

$builder->create_build_script();

if ($builder->check_installed_tidyp()) {
  my $ans = $builder->prompt("Tidyp detected, wanna use tidyp already installed on your system?", 'y');
  $builder->notes('installed_tidyp', undef) unless (lc($ans) eq 'y');
}

# delete build marker (= force rebuild after running Build.PL)
unlink 'build_done';

# set what tidyp tarball we are gonna use
$builder->notes('tidyp_url', 'http://github.com/downloads/petdance/tidyp/tidyp-1.04.tar.gz');
$builder->notes('tidyp_dir', 'tidyp-1.04'); # dirname inside the tarball
$builder->notes('tidyp_sha1', '5d9050512259c3a67a2f48469555932e3a7b8bd0'); # SHA1 checksum of the tarball


inc/My/Builder.pm
package My::Builder;

use strict;
use warnings;
use base 'Module::Build';

use lib "inc";
use File::Spec::Functions qw(catfile);
use ExtUtils::Command;
use File::Fetch;
use File::Temp qw(tempdir tempfile);
use File::Path qw();
use File::ShareDir;
use Digest::SHA qw(sha1_hex);
use Archive::Extract;
use Config;
use ExtUtils::Liblist;

sub ACTION_install {
  my $self = shift;
  my $sharedir = eval {File::ShareDir::dist_dir('Alien-Tidyp')} || '';
 
  if ( -d $sharedir ) {
    print STDERR "Removing the old '$sharedir'\n";
    File::Path::rmtree($sharedir);
    File::Path::mkpath($sharedir);    
  }
 
  return $self->SUPER::ACTION_install(@_);
}


sub ACTION_code {
  my $self = shift;

  unless (-e 'build_done') {
    $self->add_to_cleanup('build_done');
    my $inst = $self->notes('installed_tidyp');
    if (defined $inst) {
      $self->config_data('config', { LIBS   => $inst->{lflags},
                                     INC    => $inst->{cflags},
                                   });
    }
    else {
      # important directories
      my $download = 'download';
      my $patches = 'patches';
      my $build_src = 'build_src';
      # we are deriving the subdir name from VERSION as we want to prevent
      # troubles when user reinstalls the newer version of Alien::Tidyp
      my $build_out = catfile('sharedir', $self->{properties}->{dist_version});
      $self->add_to_cleanup($build_out);
      $self->add_to_cleanup($build_src);

      # get sources
      my $tarball = $self->args('srctarball');
      my $dir  = $self->notes('tidyp_dir');
      my $sha1 = $self->notes('tidyp_sha1');
      $self->notes('tidyp_src', "$build_src/$dir");      
      my $archive;
      if ($tarball && -f $tarball) {
        $archive = $tarball;
      }      
      else {
        warn "Downloading from explicit URL: '$tarball'\n" if $tarball;
        my $url  = $tarball || $self->notes('tidyp_url');
        $self->fetch_file($url, $sha1, $download);
        $archive = catfile($download, File::Fetch->new(uri => $url)->file);
      }
      print STDERR "Checking checksum '$archive'...\n";
      die "###ERROR### Checksum failed '$archive'" unless $self->check_sha1sum($archive, $sha1);
      my $ae = Archive::Extract->new( archive => $archive );
      die "###ERROR### Cannot extract tarball ", $ae->error unless $ae->extract(to => $build_src);
      die "###ERROR### Cannot find expected dir='",$self->notes('tidyp_src'),"'" unless -d $self->notes('tidyp_src');
      
      # go for build
      $self->build_binaries($build_out, $self->notes('tidyp_src'));
      # store info about build into future Alien::Tidyp::ConfigData
      $self->config_data('share_subdir', $self->{properties}->{dist_version});
      $self->config_data('config', { PREFIX => '@PrEfIx@',
                                     LIBS   => '-L' . $self->quote_literal('@PrEfIx@/lib') . ' -ltidyp',
                                     INC    => '-I' . $self->quote_literal('@PrEfIx@/include/tidyp'),
                                   });
    }
    # mark sucessfully finished build
    local @ARGV = ('build_done');
    ExtUtils::Command::touch();
  }
  $self->SUPER::ACTION_code;
}

sub fetch_file {
  my ($self, $url, $sha1sum, $download) = @_;
  die "###ERROR### _fetch_file undefined url\n" unless $url;
  die "###ERROR### _fetch_file undefined sha1sum\n" unless $sha1sum;
  my $ff = File::Fetch->new(uri => $url);
  my $fn = catfile($download, $ff->file);
  if (-e $fn) {
    print STDERR "Checking checksum for already existing '$fn'...\n";
    return 1 if $self->check_sha1sum($fn, $sha1sum);
    unlink $fn; #exists but wrong checksum
  }
  print STDERR "Fetching '$url'...\n";
  my $fullpath = $ff->fetch(to => $download);
  die "###ERROR### Unable to fetch '$url'" unless $fullpath;  
  die "###ERROR### _fetch_file failed '$fn'" unless -e $fn;
  return 1
}

sub check_sha1sum {
  my ($self, $file, $sha1sum) = @_;
  my $sha1 = Digest::SHA->new;
  my $fh;
  open($fh, $file) or die "###ERROR## Cannot check checksum for '$file'\n";
  binmode($fh);
  $sha1->addfile($fh);
  close($fh);  
  my $rv = ($sha1->hexdigest eq $sha1sum) ? 1 : 0;
  warn "###WARN## sha1 mismatch: got      '", $sha1->hexdigest , "'\n",
       "###WARN## sha1 mismatch: expected '", $sha1sum, "'\n",
       "###WARN## sha1 mismatch: filesize '", (-s $file), "'\n", unless $rv;
  return $rv;
}

sub build_binaries {
  die "###ERROR### My::Builder cannot build libtidyp from sources, use rather My::Builder::<platform>";
}

sub quote_literal {
  # this needs to be overriden in My::Builder::<platform>
  my ($self, $path) = @_;
  return $path;
}

sub check_installed_tidyp {
  my ($self) = @_;

  require ExtUtils::CBuilder;
  my $cb = ExtUtils::CBuilder->new( quiet => 1 );
  my $dir = tempdir( CLEANUP => 1 );
  my ($fs, $src) = tempfile( DIR => $dir, SUFFIX => '.c' );
  syswrite($fs, <<MARKER); # write test source code
#include <tidyp.h>
int main() { tidyVersion(); return 0; }

MARKER
  close($fs);

  my $tdir = $ENV{TIDYP_DIR} || '';
  my @candidates;
  push(@candidates, { L => "$tdir/lib", I => "$tdir/include/tidyp" }) if -d $tdir;
  push(@candidates, { L => '/usr/local/lib', I => '/usr/local/include/tidyp' });
  push(@candidates, { L => '/usr/lib', I => '/usr/include/tidyp' });
  push(@candidates, { L => '', I => "$Config{usrinc}/tidyp" });

  print STDERR "Gonna detect tidyp already installed on your system:\n";
  foreach my $i (@candidates) {
    my $lflags = $i->{L} ? '-L'.$self->quote_literal($i->{L}).' -ltidyp' : '-ltidyp';
    my $cflags = $i->{I} ? '-I'.$self->quote_literal($i->{I}) : '';
    print STDERR "- testing: $cflags $lflags ...\n";
    $lflags = ExtUtils::Liblist->ext($lflags) if($Config{make} =~ /nmake/ && $Config{cc} =~ /cl/); # MSVC compiler hack
    my ($obj, $exe);
    open(my $olderr, '>&', STDERR);
    open(STDERR, '>', File::Spec->devnull());
    $obj = eval { $cb->compile( source => $src, extra_compiler_flags => $cflags ) };
    $exe = eval { $cb->link_executable( objects => $obj, extra_linker_flags => $lflags ) } if $obj;
    open(STDERR, '>&', $olderr);
    next unless $exe;
    print STDERR "- TIDYP FOUND!\n";
    $self->notes('installed_tidyp', { lflags => $lflags, cflags => $cflags } );
    return 1;
  }
  print STDERR "- tidyp not found (we have to build it from sources)!\n";
  return 0;
}

1;


lib/Alien/Libtidyp.pm
package Alien::Libtidyp;

use warnings;
use strict;

use Alien::Libtidyp::ConfigData;
use File::ShareDir qw(dist_dir);
use File::Spec::Functions qw(catdir catfile rel2abs);

=head1 NAME

Alien::Libtidyp - building and using libtidyp binaries

=head1 VERSION

Version 1.0

=cut

our $VERSION = '1.0';

=head1 SYNOPSIS

Alien::Libtidyp during its installation does the following:

=over

=item * Builds 'libtidyp' binaries from source codes

=item * Installs binaries into 'share' directory of Alien::Libtidyp distribution

=back

Later you can use Alien::Libtidyp in your module that needs to link with libtidyp
like this:

    # Sample Makefile.pl
    use ExtUtils::MakeMaker;
    use Alien::Libtidyp;

    WriteMakefile(
      NAME         => 'Any::Libtidyp::Module',
      VERSION_FROM => 'lib/Any/Libtidyp/Module.pm',
      LIBS         => Alien::Libtidyp->config('LIBS'),
      INC          => Alien::Libtidyp->config('INC'),
      # + additional params
    );

=head1 METHODS

=head2 config()

This function is the main public interface to this module.

    Alien::Libtidyp->config('LIBS');

Returns a string like: '-L/path/to/libtidy/dir/lib -ltidyp'

    Alien::Libtidyp->config('INC');

Returns a string like: '-I/path/to/libtidy/dir/include'

    Alien::Libtidyp->config('PREFIX');

Returns a string like: '/path/to/libtidy/dir'

=head1 AUTHOR

KMX, C<< <kmx at cpan.org> >>

=head1 BUGS

Please report any bugs or feature requests to C<bug-alien-libtidyp at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Alien-Libtidyp>. 

=head1 LICENSE AND COPYRIGHT

Copyright 2010 KMX.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=cut

sub config
{
  my ($package, $param) = @_;
  my $share_dir = dist_dir('Alien-Libtidyp');
  my $subdir = Alien::Libtidyp::ConfigData->config('share_subdir');
  return unless $subdir;
  my $real_prefix = catdir($share_dir, $subdir);
  return unless ($param =~ /[a-z0-9_]*/i);
  my $val = Alien::Libtidyp::ConfigData->config('config')->{$param};
  return unless $val;  
  $val =~ s/\@PrEfIx\@/$real_prefix/g; # handle @PrEfIx@ replacement
  return $val;
}


1;


inc/My/Builder/Unix.pm
package My::Builder::Unix;

use strict;
use warnings;
use base 'My::Builder';

use File::Spec::Functions qw(catdir catfile rel2abs);
use File::Spec;
use Config;

sub build_binaries {
  my ($self, $build_out, $srcdir) = @_;
  $srcdir ||= 'src';
  my $prefixdir = rel2abs($build_out);
  $self->config_data('build_prefix', $prefixdir); # save it for future ConfigData
  
  #hack - use updated config.guess + config.sub - remove when fixed in tidyp distribution
  require File::Copy;
  File::Copy::copy("patches/config.sub","$srcdir/config.sub");
  File::Copy::copy("patches/config.guess","$srcdir/config.guess");
  #end of hack

  chdir $srcdir;

  # do './configure ...'
  my $run_configure = 'y';
  $run_configure = $self->prompt("Run ./configure again? (y/n)", "n") if (-f "config.status");
  if (lc($run_configure) eq 'y') {
    my @cmd = ( './configure', '--enable-shared=no', '--disable-dependency-tracking', "--prefix=$prefixdir");
    if ($^O eq 'darwin') {
      #this is fix for https://rt.cpan.org/Ticket/Display.html?id=66382
      push @cmd, "CFLAGS=$Config{ccflags} -fPIC";
      push @cmd, "LDFLAGS=$Config{ldflags}";
    }
    else {
      #FIXME maybe use %Config values for all UNIX systems (not now, maybe in the future)
      push @cmd, 'CFLAGS=-fPIC';
    }    
    #On solaris, some tools like 'ar' are not in the default PATH, but in /usr/???/bin
    #see failure http://www.cpantesters.org/cpan/report/138b45f2-4b6f-11e0-afaf-8138785ebe45
    if ($^O eq 'solaris' && system('ar -V') < 0) {    
      for (qw[/usr/ccs/bin /usr/xpg4/bin /usr/sfw/bin /usr/xpg6/bin /usr/gnu/bin /opt/gnu/bin /usr/bin]) {
        if (-x "$_/ar") {
          push @cmd, "AR=$_/ar";
          last;
        }
      }
    }
    print STDERR "Configuring ...\n";
    print STDERR "(cmd: ".join(' ',@cmd).")\n";
    $self->do_system(@cmd) or die "###ERROR### [$?] during ./configure ... ";
  }

  # do 'make install'
  my @cmd = ($self->get_make, 'install');
  print STDERR "Running make install ...\n";
  print STDERR "(cmd: ".join(' ',@cmd).")\n";
  $self->do_system(@cmd) or die "###ERROR### [$?] during make ... ";

  chdir $self->base_dir();
  return 1;
}

sub get_make {
  my ($self) = @_;
  my $devnull = File::Spec->devnull();
  my @try = ($Config{gmake}, 'gmake', 'make', $Config{make});
  my %tested;
  print STDERR "Gonna detect GNU make:\n";

  if ($^O eq 'cygwin') {
    print STDERR "- on cygwin always 'make'\n";
    return 'make'
  }

  foreach my $name ( @try ) {
    next unless $name;
    next if $tested{$name};
    $tested{$name} = 1;
    print STDERR "- testing: '$name'\n";
    my $ver = `$name --version 2> $devnull`;
    if ($ver =~ /GNU Make/i) {
      print STDERR "- found: '$name'\n";
      return $name
    }
  }
  print STDERR "- fallback to: 'make'\n";
  return 'make';
}

# we are not afraid of dirnames with spaces on UNIX
# IMPORTANT: EU::MM is not properly handling -L'/path/to/lib/dir'

#sub quote_literal {
#    my ($self, $txt) = @_;
#    $txt =~ s|'|'\\''|g;
#    return "'$txt'";
#}

1;


inc/My/Builder/Windows.pm
package My::Builder::Windows;

use strict;
use warnings;
use base 'My::Builder';

use File::Spec::Functions qw(catdir catfile rel2abs);
use Config;

sub build_binaries {
  my ($self, $build_out, $srcdir) = @_;
  my $prefixdir = rel2abs($build_out);
  my $perl = $^X;
  # for GNU make on MS Windows it is safer to convert \ to /
  $perl =~ s|\\|/|g;
  $prefixdir =~ s|\\|/|g;

  my $make = $self->get_make;
  print STDERR "Gonna call make install ...\n";
  my @cmd;
  if($make =~ /nmake/ && $Config{make} =~ /nmake/ && $Config{cc} =~ /cl/) { # MSVC compiler
    my $makefile = rel2abs('patches\Makefile.nmake');
    if ($Config{archname} =~ /x64/) { #64bit
      @cmd = ( $make, '-f', $makefile, "PERL=perl", "PREFIX=$prefixdir", "CFG=Win64", "install" );
    }
    else { #32bit
      @cmd = ( $make, '-f', $makefile, "PERL=perl", "PREFIX=$prefixdir", "install" );
    }    
  }
  else { # gcc compiler
    my $makefile = rel2abs('patches\Makefile.mingw');
    @cmd = ( $make, '-f', $makefile, "PERL=$perl", "PREFIX=$prefixdir", "CC=$Config{cc}", "install" );
  }
  print STDERR "[cmd: ".join(' ',@cmd)."]\n";
  chdir $srcdir;
  $self->do_system(@cmd) or die "###ERROR### [$?] during make ... ";
  chdir $self->base_dir();

  return 1;
}

sub get_make {
  my ($self) = @_;
  my @try = ( 'dmake', 'mingw32-make', 'gmake', 'make', $Config{make}, $Config{gmake} );
  print STDERR "Gonna detect make:\n";
  foreach my $name ( @try ) {
    next unless $name;
    print STDERR "- testing: '$name'\n";
    if (system("$name --help 2>nul 1>nul") != 256) {
      # I am not sure if this is the right way to detect non existing executable
      # but it seems to work on MS Windows (more or less)
      print STDERR "- found: '$name'\n";
      return $name;
    };
  }
  print STDERR "- fallback to: 'dmake'\n";
  return 'dmake';
}

sub quote_literal {
    my ($self, $txt) = @_;
    $txt =~ s|"|\\"|g;
    return qq("$txt");
}

1;


lib/Alien/Tidyp.pm
package Alien::Tidyp;

use warnings;
use strict;

use Alien::Tidyp::ConfigData;
use File::ShareDir qw(dist_dir);
use File::Spec::Functions qw(catdir catfile rel2abs);

=head1 NAME

Alien::Tidyp - Building, finding and using tidyp library - L<http://www.tidyp.com>

=cut

our $VERSION = 'v1.4.5';

=head1 VERSION

Version 1.4.5 of Alien::Tidyp uses I<tidyp> sources v1.04.

=head1 SYNOPSIS

B<IMPORTANT:> This module is not a perl binding for I<tidyp> library; it is just a helper module.
The real perl binding is implemented by e.g L<HTML::Tidy|HTML::Tidy> module, which is able to
use Alien::Tidyp to locate I<tidyp> library on your system (or build it from source codes).

Alien::Tidyp tries (in given order) during its installation:

=over

=item * Locate an already installed I<tidyp> + ask user whether to use the already installed
I<tidyp> or whether to build I<tidyp> from sources

=item * Via env variable TIDYP_DIR you can specify where the build script should look
for the already installed I<tidyp> (directories $TIDYP_DIR/lib and $TIDYP_DIR/include/tidyp
are expected to exist)

=item * When not using the already installed I<tidyp> build process continues with
the following steps

=item * Download I<tidyp> source code tarball

=item * Build I<tidyp> binaries from source codes (note: static libraries are build in this case)

=item * Install binaries and dev files (*.h, *.a) into I<share> directory of Alien::Tidyp
distribution - I<share> directory is usually something like this: /usr/lib/perl5/site_perl/5.10/auto/share/dist/Alien-Tidyp

=back

Later on you can use Alien::Tidyp in your module that needs to link with
I<tidyp> like this:

 # Sample Makefile.pl
 use ExtUtils::MakeMaker;
 use Alien::Tidyp;

 WriteMakefile(
   NAME         => 'Any::Tidyp::Module',
   VERSION_FROM => 'lib/Any/Tidyp/Module.pm',
   LIBS         => Alien::Tidyp->config('LIBS'),
   INC          => Alien::Tidyp->config('INC'),
   # + additional params
 );

B<IMPORTANT:> As Alien::Tidyp builds static libraries the modules using Alien::Tidyp (e.g. L<HTML::Tidy|HTML::Tidy>)
need to have Alien::Tidyp just for building, not for later use. In other words Alien:Tidyp is just "build dependency"
not "run-time dependency".

=head1 Build.PL options

=head2 --srctarball=<url_or_filename>

This option might come handy if you are not connected to the Internet. You can use it like:

 Build.PL --srctarball=/path/to/file/tidyp-1.04.tar.gz

or:

 Build.PL --srctarball=http://any.server.com/path/to/tidyp-1.04.tar.gz

IMPORTANT: The file should always be exactly the same source code tarball as specified in the end of Build.PL - see
source code (SHA1 checksum of the tarball is checked).
 
=head1 METHODS

=head2 config()

This function is the main public interface to this module.

 Alien::Tidyp->config('LIBS');

Returns a string like: '-L/path/to/tidyp/dir/lib -ltidyp'

 Alien::Tidyp->config('INC');

Returns a string like: '-I/path/to/tidyp/dir/include/tidyp'

 Alien::Tidyp->config('PREFIX');

Returns a string like: '/path/to/tidyp/dir' (note: if using the already installed
tidyp config('PREFIX') returns undef)

=head1 AUTHOR

KMX, E<lt>kmx at cpan.orgE<gt>

=head1 BUGS

Please report any bugs or feature requests to E<lt>bug-Alien-Tidyp at rt.cpan.orgE<gt>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Alien-Tidyp>.

=head1 LICENSE AND COPYRIGHT

Please notice that the source code of tidyp library has a different license than module itself.

=head2 Alien::Tidyp perl module

Copyright (c) 2010 KMX.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.

=head2 Source code of tidyp library

Copyright (c) 1998-2003 World Wide Web Consortium
(Massachusetts Institute of Technology, European Research
Consortium for Informatics and Mathematics, Keio University).
All Rights Reserved.

This software and documentation is provided "as is," and
the copyright holders and contributing author(s) make no
representations or warranties, express or implied, including
but not limited to, warranties of merchantability or fitness
for any particular purpose or that the use of the software or
documentation will not infringe any third party patents,
copyrights, trademarks or other rights.

The copyright holders and contributing author(s) will not be held
liable for any direct, indirect, special or consequential damages
arising out of any use of the software or documentation, even if
advised of the possibility of such damage.

Permission is hereby granted to use, copy, modify, and distribute
this source code, or portions hereof, documentation and executables,
for any purpose, without fee, subject to the following restrictions:

1. The origin of this source code must not be misrepresented.

2. Altered versions must be plainly marked as such and must not be misrepresented as being the original source.

3. This Copyright notice may not be removed or altered from any  source or altered source distribution.

The copyright holders and contributing author(s) specifically
permit, without fee, and encourage the use of this source code
as a component for supporting the Hypertext Markup Language in
commercial products. If you use this source code in a product,
acknowledgment is not required but would be appreciated.

=cut

sub config
{
  my ($package, $param) = @_;
  return unless ($param =~ /[a-z0-9_]*/i);
  my $subdir = Alien::Tidyp::ConfigData->config('share_subdir');
  unless ($subdir) {
    #we are using tidyp already installed on your system not compiled by Alien::Tidyp
    #therefore no additinal magic needed
    return Alien::Tidyp::ConfigData->config('config')->{$param};
  }
  my $share_dir = dist_dir('Alien-Tidyp');
  my $real_prefix = catdir($share_dir, $subdir);
  my $val = Alien::Tidyp::ConfigData->config('config')->{$param};
  return unless $val;
  $val =~ s/\@PrEfIx\@/$real_prefix/g; # handle @PrEfIx@ replacement
  return $val;
}

1;
